src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleLeaderElection.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.amazonaws.services.kinesis.clientlibrary.lib.worker;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.amazonaws.services.kinesis.clientlibrary.lib.periodicshardsync.LeaderElectionStrategy;
+import com.amazonaws.services.kinesis.leases.exceptions.DependencyException;
+import com.amazonaws.services.kinesis.leases.exceptions.InvalidStateException;
+import com.amazonaws.services.kinesis.leases.exceptions.ProvisionedThroughputException;
+import com.amazonaws.services.kinesis.leases.impl.KinesisClientLease;
+import com.amazonaws.services.kinesis.leases.impl.LeaseManager;
+import com.amazonaws.services.kinesis.leases.interfaces.ILeaseManager;
+import com.amazonaws.util.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * This implementation of the {@code LeaderElectionStrategy} elects the leaders
+ * by firstly getting all the leases. The leases are then shuffled using a
+ * predetermined constant seed so that the lease ordering is preserved across
+ * workers. This strategy helps in randomly shuffling the leases in order to
+ * reduce the probability of choosing the leader workers co-located on the same
+ * host in case the lease owner field in the lease doesn't distinguish between workers
+ * on the same host e.g using ip address of the host for the lease owner for all workers
+ * on a host. The number of leaders to elect is configurable here
+ * {@code config.getPeriodicShardSyncMaxWorkers()}. Once the leaders are
+ * elected, all the registered listeners are informed about the new election
+ */
+class DeterministicShuffleLeaderElection
+        implements LeaderElectionStrategy<KinesisClientLease>, Comparator<KinesisClientLease> {
+
+    static final String LEAES_CACHE_KEY = "all_leases";
+    // Fixed seed so that the shuffle order is preserved across workers
+    static final int DETERMINISTIC_SHUFFLE_SEED = 1947;
+    private static final int PERIODIC_SHARD_SYNC_MAX_WORKERS = 5;
```

Comment:
How will this be set differently for Kinesis and DDB? Kinesis wants a single worker.

Reply:
The default value of PERIODIC_SHARD_SYNC_MAX_WORKERS here would be 1 and the ddb streams adapter would inject another implementation of the LeaderElectionStrategy with number of leader workers as 5